contracts/token/ERC20WithPermit.sol

```diff
@@ -2,6 +2,224 @@
 
 pragma solidity 0.8.4;
 
-contract Erc20WithPermit {
-    // TODO will be implemented separately
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./IERC20WithPermit.sol";
+import "./IReceiveApproval.sol";
+
+/// @title  ERC20WithPermit
+/// @notice Burnable ERC20 token with EIP2612 permit functionality. User can
+///         authorize a transfer of their token with a signature conforming
+///         EIP712 standard instead of an on-chain transaction from their
+///         address. Anyone can submit this signature on the user's behalf by
+///         calling the permit function, as specified in EIP2612 standard,
+///         paying gas fees, and possibly performing other actions in the same
+///         transaction.
```

Comment:
Don't we want to implement `IERC20Metadata` as OZ's `ERC20` does?

Reply:
See 6cba00526a60ba195d252e34c200d2be868b463b please.